src/Tasks/Delete.cs

```diff
@@ -60,43 +93,58 @@ public void Cancel()
         /// </summary>
         public override bool Execute()
         {
+            if (!ValidateInputs())
+            {
+                return false;
+            }
             var deletedFilesList = new List<ITaskItem>();
             var deletedFilesSet = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
 
             foreach (ITaskItem file in Files)
             {
-                if (_canceling)
-                {
-                    return false;
-                }
-
-                try
+                int retries = 0;
+                while (!_canceling)
```

Comment:
I have some reservations about this cycle. 
The main one is that in case of cancellation we do not return from this function as before but have to run the foreach cycle till its end. 
The minor one, we have two more conditions when we want out of the cycle - when we success or when we exceed the number of retries. For better code understanding I would wish to read those conditions here instead of searching below where we break out of the infinite cycle (unless there is a good reason not to, of course).

Reply:
I'm a little confused by your first reservation. If they cancel before we get into the loop, we'd skip over it as before. If not, we'd stop right after the attempt when they cancel; that should be equivalent to what it was before when we'd go to the end of the task execution if they cancel after it started.

Reply:
> For better code understanding I would wish to read those conditions here instead of searching below where we break out of the infinite cycle (unless there is a good reason not to, of course).

Add comment

Reply:
> I'm a little confused by your first reservation. If they cancel before we get into the loop, we'd skip over it as before. If not, we'd stop right after the attempt when they cancel; that should be equivalent to what it was before when we'd go to the end of the task execution if they cancel after it started.

There was one (foreach) cycle before, from which we return in case of cancellation. Now we have two (foreach and while) cycles, and we skip only internal one in case of cancellation.

Reply:
> > For better code understanding I would wish to read those conditions here instead of searching below where we break out of the infinite cycle (unless there is a good reason not to, of course).
> 
> Add comment

I would rather have code self-explanatory here. The breaks are in the end of the cycle, which makes me think that cycle could be written without them. But as I said, it is a minor thing.

Reply:
> > I'm a little confused by your first reservation. If they cancel before we get into the loop, we'd skip over it as before. If not, we'd stop right after the attempt when they cancel; that should be equivalent to what it was before when we'd go to the end of the task execution if they cancel after it started.
> 
> There was one (foreach) cycle before, from which we return in case of cancellation. Now we have two (foreach and while) cycles, and we skip only internal one in case of cancellation.

Yeah, it's a listtle different from before. But I think it's resonable. If cancel the delete operation, it will return false directly before. But now it will output the deletedfiles we have done before cancel and return !Log.HasLoggedErrors. I think that is more friendly for users.

Reply:
Well, the cancellation I think should cancel the operation as soon as possible, which is not in this new code. Imagine that somebody by mistake put millions of files there. And wants to cancel. It could be very long wait. And it is still possible to return the already deleted files when cancelled, nothing prevents that.

Reply:
Thanks @AR-May.  Updated and please help review again.

Reply:
We can move the cancellation inside the "while" cycle, and it will work a bit better in case of cancellation but it is worse for other cases, so I am not sure about that.  I think current version is fine.

Reply:
Put cancellation inside the while cycle，it's better when Retries is large. But put cancellation outside the while cycle，it's better when a lot of dupulicate files input